projects/pgai/pgai/alembic/configuration.py

```diff
@@ -0,0 +1,227 @@
+import re
+from dataclasses import dataclass, fields
+from datetime import timedelta
+from typing import ClassVar, Literal, Protocol, runtime_checkable
+
+from pydantic import BaseModel
+
+from pgai.vectorizer.base import (
+    BaseOllamaConfig,
+    BaseOpenAIConfig,
+    BaseProcessing,
+    BasePythonTemplate,
+    BaseVoyageAIConfig,
+    ChunkingCharacterTextSplitter,
+    ChunkingRecursiveCharacterTextSplitter,
+)
+
+
+@runtime_checkable
+class SQLArgumentProvider(Protocol):
+    def to_sql_argument(self) -> str: ...
+
+
+@runtime_checkable
+class PythonArgProvider(Protocol):
+    def to_python_arg(self) -> str: ...
+
+
+def format_sql_params(params: dict[str, str | None | bool | list[str]]) -> str:
+    """Format dictionary of parameters into SQL argument string without any quoting."""
+    formatted: list[str] = []
+    for key, value in params.items():
+        if value is None:
+            continue
+        elif isinstance(value, bool):
+            formatted.append(f"{key}=>{str(value).lower()}")
+        elif isinstance(value, list):
+            array_list = ",".join(f"E'{v}'" for v in value)
+            formatted.append(f"{key}=>ARRAY[{array_list}]")
+        elif isinstance(value, timedelta):
+            formatted.append(f"{key}=>'{value.seconds} seconds'")
+        else:
+            formatted.append(f"{key}=> '{value}'")
+    return ", ".join(formatted)
+
+
+class SQLArgumentMixin:
+    arg_type: ClassVar[str]
+    function_name: ClassVar[str | None] = None
+
+    def to_sql_argument(self) -> str:
+        # Get all fields including from parent classes
+        params = {}
+        for field_name, _field in self.model_fields.items():  # type: ignore
+            if field_name != "arg_type":
```

Comment:
is the function_name field included then, how does that work?

Reply:
not sure how that worked, but now its excluded too :D